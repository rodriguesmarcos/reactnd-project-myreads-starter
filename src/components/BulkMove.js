import React from 'react';
import PropTypes from 'prop-types';
import serializeForm from 'form-serialize';

const BulkMove = ({children, bar, onBulkMove, shelves, books, showing, shelf = 'none'}) => {

  const onSubmit = handleMoveBooks;
  const select = renderBulkMoveSelect(shelves, shelf, onBulkMove);

  if ( showing.length === 0) {
    return children;
  }

  const childrenWithProps = React.Children.map(children,
    (child) => React.cloneElement(child, {
      isMovable: true
    })
  );

  return (
    <form onSubmit={(e)=> onSubmit(e, onBulkMove, books)}>
      <div>
        Move Books: { select }
        <button type="submit">Move</button>
      </div>

      {childrenWithProps}
    </form>
  )
};

const renderBulkMoveSelect = (shelves, shelf, onBulkMove) => {
  const noneOpt = [{ id: 'none', name: 'None' }];
  let shelvesOptions = shelves;
  if ( shelf.id )
    shelvesOptions = shelves.filter(s => s.id !== shelf.id ).concat(noneOpt);

  return (
    <select
      name="shelf"
      defaultValue=""
      onSubmit={onBulkMove}
    >
      <option
        value=""
        disabled={true}>
      Choose a shelf</option>
      {shelvesOptions.map(o => (
        <option
          key={o.id}
          value={o.id}>
        {o.name}</option>
      ))}
    </select>
  )

}

const handleMoveBooks = (e, onBulkMove, books) => {
  e.preventDefault();
  const values = serializeForm( e.target, { 'hash': true } );
  const { bookId:booksIds, shelf } = values;
  const changeBooks = books.filter( b => booksIds.includes(b.id)).filter( b => shelf !== b.shelf);

  onBulkMove( changeBooks, shelf );
}

BulkMove.propTypes = {
  bar: PropTypes.func,
  books: PropTypes.array.isRequired,
}

export default BulkMove;
