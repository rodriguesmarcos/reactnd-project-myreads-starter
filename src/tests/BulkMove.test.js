import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import BulkMove from '../components/BulkMove.js';
import books from './fixtures/books';
import shelves from './fixtures/shelves';

describe('<BulkMove />', () => {

  const minProps = {
    books: books,
    shelves: shelves,
    showing: [],
    onBulkMove: () => {},
  }

  test('renders without exploding', () => {
    const wrapper = shallow(<BulkMove {...minProps}><div></div></BulkMove>);
    expect(wrapper.length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render children', () => {
    const wrapper = shallow(<BulkMove {...minProps}><div></div></BulkMove>);
    expect(wrapper.find('div').length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render children', () => {
    const wrapper = shallow(<BulkMove {...minProps} showing={books}><div></div></BulkMove>);
    expect(wrapper.find('form').length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should call onBulkMove', () => {
    const onBulkMoveSpy = jest.fn();
    const value = shelves[0].id;
    const wrapper = shallow(<BulkMove {...minProps} showing={books} onBulkMove={onBulkMoveSpy}><div></div></BulkMove>);
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    });
    expect(onBulkMoveSpy).toHaveBeenCalledWith([], undefined);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
