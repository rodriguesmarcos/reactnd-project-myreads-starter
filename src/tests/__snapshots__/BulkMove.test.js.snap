exports[`<BulkMove /> renders without exploding 1`] = `<div />`;

exports[`<BulkMove /> should call onBulkMove 1`] = `
<form
  onSubmit={[Function]}>
  <div>
    Move Books: 
    <select
      defaultValue=""
      name="shelf"
      onSubmit={[Function]}>
      <option
        disabled={true}
        value="">
        Choose a shelf
      </option>
      <option
        key="currentlyReading"
        value="currentlyReading">
        Currently Reading
      </option>
      <option
        key="wantToRead"
        value="wantToRead">
        Want to Read
      </option>
      <option
        key="read"
        value="read">
        Read
      </option>
    </select>
    <button
      type="submit">
      Move
    </button>
  </div>
  <div
    isMovable={true}
    key=".0" />
</form>
`;

exports[`<BulkMove /> should render children 1`] = `<div />`;

exports[`<BulkMove /> should render children 2`] = `
<form
  onSubmit={[Function]}>
  <div>
    Move Books: 
    <select
      defaultValue=""
      name="shelf"
      onSubmit={[Function]}>
      <option
        disabled={true}
        value="">
        Choose a shelf
      </option>
      <option
        key="currentlyReading"
        value="currentlyReading">
        Currently Reading
      </option>
      <option
        key="wantToRead"
        value="wantToRead">
        Want to Read
      </option>
      <option
        key="read"
        value="read">
        Read
      </option>
    </select>
    <button
      type="submit">
      Move
    </button>
  </div>
  <div
    isMovable={true}
    key=".0" />
</form>
`;
